### PR TITLE
Use money64 (from NSF)

### PR DESCRIPTION
**Pull request by Gymnasiast**

### CODE REVIEW

File: src/openrct2/scenario/Scenario.h [hunk 1]
```diff
@@ -423,7 +423,7 @@ enum
 #define AUTOSAVE_PAUSE 0
 #define DEFAULT_NUM_AUTOSAVES_TO_KEEP 10
 
-static constexpr money32 COMPANY_VALUE_ON_FAILED_OBJECTIVE = 0x80000001;
+static constexpr money64 COMPANY_VALUE_ON_FAILED_OBJECTIVE = 0x8000000000000001;
```
Comment (Gymnasiast) [inline src/openrct2/scenario/Scenario.h:426]: It needs to be checked if old saves that failed the objective are imported correctly.

File: src/openrct2/rct2/S6Importer.cpp [hunk 2]
```diff
@@ -230,8 +230,8 @@ class S6Importer final : public IParkImporter
         ImportTileElements();
         ImportEntities();
 
-        gInitialCash = _s6.initial_cash;
-        gBankLoan = _s6.current_loan;
+        gInitialCash = ToMoney64(_s6.initial_cash);
```
Comment (Gymnasiast) [inline src/openrct2/rct2/S6Importer.cpp:233]: All conversions need porting to the S6Exporter as well.

File: src/openrct2/common.h [hunk 3]
```diff
@@ -141,6 +141,12 @@ using money64 = fixed64_1dp;
 #define MONEY_FREE MONEY(0, 00)
 #define MONEY16_UNDEFINED static_cast<money16>(static_cast<uint16_t>(0xFFFF))
 #define MONEY32_UNDEFINED (static_cast<money32>(0x80000000))
+#define MONEY64_UNDEFINED (static_cast<money64>(0x8000000000000000))
+
+static constexpr money64 ToMoney64(money32 value)
+{
+    return value == MONEY32_UNDEFINED ? MONEY64_UNDEFINED : value;
+}
```
Comment (Gymnasiast) [inline src/openrct2/common.h:149]: Check if we need a variant for money16 as well.

We definitely need reverse versions of this function.

**Comment by duncanspumpkin** [comment]
New replays required for the moneyeffect change this is causing the test failure. I'm just recording new ones atm

File: src/openrct2/rct1/S4Importer.cpp [hunk 4]
```diff
@@ -2300,7 +2300,7 @@ class S4Importer final : public IParkImporter
         // This is corrected here, but since scenario_objective_currency doubles as minimum excitement rating,
         // we need to check the goal to avoid affecting scenarios like Volcania.
         if (_s4.scenario_objective_type == OBJECTIVE_PARK_VALUE_BY)
-            gScenarioObjective.Currency = CorrectRCT1ParkValue(_s4.scenario_objective_currency);
+            gScenarioObjective.Currency = ToMoney64(CorrectRCT1ParkValue(_s4.scenario_objective_currency));
         else
             gScenarioObjective.Currency = _s4.scenario_objective_currency;
```
Comment (Gymnasiast) [inline src/openrct2/rct1/S4Importer.cpp:2305]:
```suggestion
            gScenarioObjective.Currency = ToMoney64(_s4.scenario_objective_currency);
```

File: src/openrct2/rct12/RCT12.cpp [hunk 5]
```diff
@@ -1349,3 +1350,19 @@ RCT12TrackType OpenRCT2FlatTrackTypeToRCT12(track_type_t origTrackType)
 
     return origTrackType;
 }
+
+money64 RCT12CompletedCompanyValueToOpenRCT2(money32 origValue)
```
Comment (Gymnasiast) [inline src/openrct2/rct12/RCT12.cpp:1354]: Also to constexpr

File: src/openrct2/scenario/ScenarioRepository.cpp [hunk 6]
```diff
@@ -612,7 +612,7 @@ class ScenarioRepository final : public IScenarioRepository
                 scenario_highscore_entry* highscore = InsertHighscore();
                 highscore->fileName = fs.ReadString();
                 highscore->name = fs.ReadString();
-                highscore->company_value = fs.ReadValue<money32>();
+                highscore->company_value = fileVersion == 1 ? fs.ReadValue<money32>() : fs.ReadValue<money64>();
```
Comment (duncanspumpkin) [inline src/openrct2/scenario/ScenarioRepository.cpp:615]: doesnt the money32 need to be converted? Is a money32_undefined possible or `COMPANY_VALUE_ON_FAILED_OBJECTIVE`

**Comment by duncanspumpkin** [comment]
This all looks about fine. Remember to increment network version and plugin api

File: src/openrct2/scenario/ScenarioRepository.cpp (hunk 6 above)
Comment (Gymnasiast) [reply src/openrct2/scenario/ScenarioRepository.cpp:615]: No, these are not possible. Only actual company values end up in the highscores.

**Review by duncanspumpkin — APPROVED** [review]

**Pull request merged**